File: AngularJSAuthentication.API/CRM/NetProfitController.cs
using NLog;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Security.Claims;
using System.Web.Http;

namespace AngularJSAuthentication.API.Controllers
{
    [RoutePrefix("api/NetProfit")]
    public class NetProfitController : ApiController
    {
      
        //public static Logger logger = LogManager.GetCurrentClassLogger();

        //[Route("")]
        //[HttpGet]
        //public dynamic GetDay(DateTime? datefrom, DateTime? dateto, int type, string ids)
        //{
        //    logger.Info("start OrderMaster: ");
        //    DateTime start = DateTime.Parse("01-01-2017 00:00:00");
        //    DateTime end = DateTime.Today.AddDays(1);
        //    try
        //    {
        //        var identity = User.Identity as ClaimsIdentity;
        //        int compid = 0, userid = 0;
        //        foreach (Claim claim in identity.Claims)
        //        {
        //            if (claim.Type == "compid")
        //            {
        //                compid = int.Parse(claim.Value);
        //            }
        //            if (claim.Type == "userid")
        //            {
        //                userid = int.Parse(claim.Value);
        //            }
        //        }
        //        logger.Info("User ID : {0} , Company Id : {1}", compid, userid);

        //        if (datefrom == null)
        //        {
        //            datefrom = DateTime.Parse("01-01-2017 00:00:00");
        //            dateto = DateTime.Today.AddDays(1);
        //        }
        //        else
        //        {
        //            start = datefrom.GetValueOrDefault();
        //            end = dateto.GetValueOrDefault();
        //        }
        //        var array = ids.Split(',');
        //        List<repo> report = new List<repo>();
        //        foreach (var iidd in array)
        //        {
        //            int id = Convert.ToInt32(iidd);
        //            var res = getdata(datefrom, dateto, type, id);
        //            List<netmargin> list = new List<netmargin>();
        //            List<netmargin> netData = new List<netmargin>();

        //            list = res;
        //            if (list.Count != 0)
        //            {
        //                repo r = new repo();
        //                netmargin m = new netmargin();
        //                foreach (var a in list)
        //                {

        //                    netData.Add(a);

        //                    if (a.NetPurchasePrice != 0)
        //                    {
        //                        m.SellingPrice += a.QTY * a.SellingPrice;
        //                        m.NetPurchasePrice += a.QTY * a.NetPurchasePrice;
        //                    }
        //                }
        //                if (m.SellingPrice > 0)
        //                {

        //                    try
        //                    {
        //                        r.value = (m.SellingPrice - m.NetPurchasePrice) / (m.SellingPrice) * 100;
        //                    }
        //                    catch (Exception sdf)
        //                    {

        //                    }


        //                }

        //                else
        //                {
        //                    r.value = 0;
        //                }


        //                r.netData = netData;
        //                report.Add(r);
        //            }



        //        }
        //        return report;
        //    }
        //    catch (Exception ex)
        //    {
        //        logger.Error("Error in OrderMaster " + ex.Message);
        //        logger.Info("End  OrderMaster: ");
        //        return null;
        //    }
        //}

        //public List<netmargin> getdata(DateTime? datefrom, DateTime? dateto, int type, int id)
        //{
        //    List<netmargin> list = new List<netmargin>();
        //    if (datefrom != null && dateto != null)
        //    {
        //        if (type == 4)
        //        {
        //            list = (from i in db.DbOrderDetails
        //                    join j in db.DbOrderMaster on i.OrderId equals j.OrderId
        //                    where i.CreatedDate >= datefrom && i.CreatedDate <= dateto && i.ExecutiveId == id
        //                    select new netmargin
        //                    {
        //                        QTY = i.qty,
        //                        NetPurchasePrice = i.NetPurchasePrice,
        //                        SellingPrice = i.UnitPrice,
        //                        salespersonid = i.ExecutiveId,
        //                        SalesPersonName = i.ExecutiveName,
        //                        HubName = i.WarehouseName,
        //                        CityName = i.City,
        //                        Number = i.itemNumber,
        //                        itemname = i.itemname,
        //                        SubcategoryName = i.SubcategoryName,
        //                        SubsubcategoryName = i.SubsubcategoryName,
        //                        SupplierName = i.SupplierName,
        //                        TaxPercentage = i.TaxPercentage,
        //                    }).ToList();
        //            // list = data.Where(x => x.salespersonid == id).ToList();
        //        }
        //        else if (type == 2)
        //        {
        //            list = (from i in db.DbOrderDetails
        //                    where i.CreatedDate >= datefrom && i.CreatedDate <= dateto && i.WarehouseId == id
        //                    join j in db.DbOrderMaster on i.OrderId equals j.OrderId
        //                    select new netmargin
        //                    {
        //                        QTY = i.qty,
        //                        NetPurchasePrice = i.NetPurchasePrice,
        //                        SellingPrice = i.UnitPrice,
        //                        HubId = i.WarehouseId,
        //                        HubName = i.WarehouseName,
        //                        SalesPersonName = j.SalesPerson,
        //                        CityName = i.City,
        //                        Number = i.itemNumber,
        //                        itemname = i.itemname,
        //                        SubcategoryName = i.SubcategoryName,
        //                        SubsubcategoryName = i.SubsubcategoryName,
        //                        SupplierName = i.SupplierName,
        //                        TaxPercentage = i.TaxPercentage,
        //                    }).ToList();
        //        }
        //        else if (type == 3)
        //        {
        //            list = (from i in db.DbOrderDetails
        //                    where i.CreatedDate >= datefrom && i.CreatedDate <= dateto && i.CityId == id
        //                    join j in db.DbOrderMaster on i.OrderId equals j.OrderId
        //                    select new netmargin
        //                    {
        //                        QTY = i.qty,
        //                        NetPurchasePrice = i.NetPurchasePrice,
        //                        SellingPrice = i.UnitPrice,
        //                        CityId = i.CityId,
        //                        CityName = i.City,
        //                        HubName = i.WarehouseName,
        //                        SalesPersonName = j.SalesPerson,
        //                        Number = i.itemNumber,
        //                        itemname = i.itemname,
        //                        SubcategoryName = i.SubcategoryName,
        //                        SubsubcategoryName = i.SubsubcategoryName,
        //                        SupplierName = i.SupplierName,
        //                        TaxPercentage = i.TaxPercentage,
        //                    }).ToList();
        //        }
        //        else if (type == 5)
        //        {
        //            var data = (from i in db.DbOrderDetails
        //                        where i.CreatedDate >= datefrom && i.CreatedDate <= dateto
        //                        join j in db.DbOrderMaster on i.OrderId equals j.OrderId
        //                        select new netmargin
        //                        {
        //                            QTY = i.qty,
        //                            NetPurchasePrice = i.NetPurchasePrice,
        //                            SellingPrice = i.UnitPrice,
        //                            clusterId = j.ClusterId,
        //                            CityName = i.City,
        //                            ClusterName = j.ClusterName,
        //                            HubName = i.WarehouseName,
        //                            SalesPersonName = j.SalesPerson,
        //                            Number = i.itemNumber,
        //                            itemname = i.itemname,
        //                            SubcategoryName = i.SubcategoryName,
        //                            SubsubcategoryName = i.SubsubcategoryName,
        //                            SupplierName = i.SupplierName,
        //                            TaxPercentage = i.TaxPercentage,
        //                        }).ToList();
        //            list = data.Where(x => x.clusterId == id).ToList();
        //        }
        //    }
        //    return list;
        //}
    }
    public class netmargin
    {
        public int QTY { get; set; }
        public string SubcategoryName { get; set; }
        public string SubsubcategoryName { get; set; }
        public string SupplierName { get; set; }
        public string itemname { get; set; }
        public string Number { get; set; }
        public double NetPurchasePrice { get; set; }
        public double SellingPrice { get; set; }
        public int salespersonid { get; set; }
        public int clusterId { get; set; }
        public int HubId { get; set; }
        public string HubName { get; set; }
        public int? CityId { get; set; }
        public string CityName { get; set; }
        public string ClusterName { get; set; }
        public string SalesPersonName { get; set; }
        public double? TaxPercentage { get; set; }

    }
    public class repo
    {
        public double value { get; set; }
        public List<netmargin> netData { get; set; }
    }
}

File: AngularJSAuthentication.API/CRM/NetprofitController.js
'use strict'
app.controller('NetprofitController', ['$scope', "$filter", "$http", "ngTableParams", function ($scope, $filter, $http, ngTableParams) {
    var UserRole = JSON.parse(localStorage.getItem('RolePerson'));
    //if (UserRole.role== 'HQ Master login' && UserRole.StatisticalRep == 'True' || UserRole.role== 'WH Master login' && UserRole.StatisticalRep == 'True'|| UserRole.role== 'HQ HR lead' && UserRole.StatisticalRep == 'True'|| UserRole.role== 'HQ Accounts Associate' && UserRole.StatisticalRep == 'True'|| UserRole.role== 'HQ Accounts Executive' && UserRole.StatisticalRep == 'True'|| UserRole.role== 'HQ Accounts Lead' && UserRole.StatisticalRep == 'True'|| UserRole.role== 'HQ CS Associate' && UserRole.StatisticalRep == 'True'|| UserRole.role== 'HQ CS EXECUTIVE' && UserRole.StatisticalRep == 'True'|| UserRole.role== 'HQ CS LEAD' && UserRole.StatisticalRep == 'True'|| UserRole.role== 'HQ Growth Associate' && UserRole.StatisticalRep == 'True'|| UserRole.role== 'HQ Growth Executive' && UserRole.StatisticalRep == 'True'|| UserRole.role== 'HQ Growth Lead' && UserRole.StatisticalRep == 'True'|| UserRole.role== 'HQ HR associate' && UserRole.StatisticalRep == 'True'|| UserRole.role== 'HQ HR executive' && UserRole.StatisticalRep == 'True'|| UserRole.role== 'HQ IC Associate' && UserRole.StatisticalRep == 'True'|| UserRole.role== 'HQ IC Executive' && UserRole.StatisticalRep == 'True'|| UserRole.role== 'HQ IC Lead' && UserRole.StatisticalRep == 'True'|| UserRole.role== 'HQ Marketing Associate' && UserRole.StatisticalRep == 'True'|| UserRole.role== 'HQ Marketing Executive' && UserRole.StatisticalRep == 'True'|| UserRole.role== 'HQ Marketing Lead' && UserRole.StatisticalRep == 'True'|| UserRole.role== 'HQ MIS Associate' && UserRole.StatisticalRep == 'True'|| UserRole.role== 'HQ MIS EXECUTIVE' && UserRole.StatisticalRep == 'True'|| UserRole.role== 'HQ MIS LEAD' && UserRole.StatisticalRep == 'True'|| UserRole.role== 'HQ Ops Associate' && UserRole.StatisticalRep == 'True'|| UserRole.role== 'HQ Ops Executive' && UserRole.StatisticalRep == 'True'|| UserRole.role== 'HQ Ops Lead' && UserRole.StatisticalRep == 'True'|| UserRole.role== 'HQ Purchase Associate' && UserRole.StatisticalRep == 'True'|| UserRole.role== 'HQ Purchase Executive ' && UserRole.StatisticalRep == 'True'|| UserRole.role== 'HQ Purchase Lead' && UserRole.StatisticalRep == 'True'|| UserRole.role== 'HQ Sales Associate' && UserRole.StatisticalRep == 'True'|| UserRole.role== 'HQ Sales Executive' && UserRole.StatisticalRep == 'True'|| UserRole.role== 'HQ Sales Lead' && UserRole.StatisticalRep == 'True'|| UserRole.role== 'HQ Sourcing Associate' && UserRole.StatisticalRep == 'True'|| UserRole.role== 'HQ Sourcing Executive' && UserRole.StatisticalRep == 'True'|| UserRole.role== 'HQ Sourcing Lead' && UserRole.StatisticalRep == 'True'|| UserRole.role== 'WH Agents' && UserRole.StatisticalRep == 'True'|| UserRole.role== 'WH cash manager' && UserRole.StatisticalRep == 'True'|| UserRole.role== 'WH delivery planner' && UserRole.StatisticalRep == 'True'|| UserRole.role== 'WH Ops associate' && UserRole.StatisticalRep == 'True'|| UserRole.role== 'WH Purchase executive' && UserRole.StatisticalRep == 'True'|| UserRole.role== 'WH Super visor' && UserRole.StatisticalRep == 'True'|| UserRole.role== 'WH Service lead' && UserRole.StatisticalRep == 'True'|| UserRole.role== 'WH Sales lead' && UserRole.StatisticalRep == 'True'|| UserRole.role== 'WH sales executives' && UserRole.StatisticalRep == 'True'|| UserRole.role== 'WH sales associates' && UserRole.StatisticalRep == 'True') {
       $scope.datrange = '';
        $scope.datefrom = '';
        $scope.dateto = '';
        $(function () {
            $('input[name="daterange"]').daterangepicker({
                timePicker: true,
                timePickerIncrement: 5,
                timePicker24Hour: true,
                format: 'YYYY-MM-DD h:mm A'
            });
            $('.input-group-addon').click(function () {
                $('input[name="daterange"]').trigger("select");
               // document.getElementsByClassName("daterangepicker")[0].style.display = "block";

            });
        });
        $scope.getFormatedDate = function (date) {
            console.log("here get for")
            date = new Date(date);
            var dd = date.getDate();
            var mm = date.getMonth() + 1; //January is 0!
            var yyyy = date.getFullYear();
            if (dd < 10) {
                dd = '0' + dd;
            }
            if (mm < 10) {
                mm = '0' + mm;
            }
            var date = yyyy + '-' + mm + '-' + dd;
            return date;
        }
        var Allbydate = [];
        var finallistoflist = [];
        $scope.filtype = "";
        $scope.type = "";

        var titleText = "";
        var legendText = "";
        
        $scope.selTime = [
      { value: 1, text: "Day" },
      { value: 2, text: "Month" },
      { value: 3, text: "Year" }
        ];
        $scope.selType = [
           { value: 1, text: "Total Order" },
           { value: 2, text: "Total sale" },
           { value: 3, text: "Active Retailer" },
           { value: 4, text: "Active Brand" }
        ];
        $scope.selectType = [
            { value: 4, text: "Excecutive" },
            { value: 2, text: "Hub" },
            { value: 3, text: "City" },
            { value: 5, text: "Cluster" }
        ];
        $scope.dataselect = [];
        $scope.examplemodel = [];
        $scope.exampledata = $scope.dataselect;
        $scope.examplesettings = {
            displayProp: 'DisplayName', idProp: 'PeopleID',
            scrollableHeight: '300px',
            scrollableWidth: '450px',
            enableSearch: true,
            scrollable: true
        };
        $scope.subexamplemodel = [];
        $scope.subexamplemodel = $scope.dataselect;
        $scope.subexamplesettings = {
            displayProp: 'WarehouseName', idProp: 'WarehouseId',
            scrollableHeight: '300px',
            scrollableWidth: '450px',
            enableSearch: true,
            scrollable: true
        };
        $scope.ssubexamplemodel = [];
        $scope.ssubexamplemodel = $scope.dataselect;
        $scope.ssubexamplesettings = {
            displayProp: 'CityName', idProp: 'Cityid',
            scrollableHeight: '300px',
            scrollableWidth: '450px',
            enableSearch: true,
            scrollable: true
        };
        $scope.clusterModel = [];
        $scope.clusterModel = $scope.dataselect;
        $scope.clusterSetting = {
            displayProp: 'ClusterName', idProp: 'ClusterId',
            scrollableHeight: '300px',
            scrollableWidth: '450px',
            enableSearch: true,
            scrollable: true
        };


        $scope.selectedtypeChanged = function (data) {
            $scope.dataselect = [];
            var url = serviceBase + "api/Report/select?value=" + data.type;
            $http.get(url)
            .success(function (data) {
                if (data.length == 0) {
                    alert("Not Found");
                }
                $scope.dataselect = data;
                console.log(data);
            });
        };

        alasql.fn.myfmt = function (n) {
            return Number(n).toFixed(2);
        }

        $scope.genereteReport = function (data1) {
            
            $scope.dataforsearch = { datefrom: "", dateto: "" };
            var f = $('input[name=daterangepicker_start]');
            var g = $('input[name=daterangepicker_end]');

            if (!$('#dat').val()) {
                $scope.dataforsearch.datefrom = '';
                $scope.dataforsearch.dateto = '';
            }
            else {
                $scope.dataforsearch.datefrom = f.val();
                $scope.dataforsearch.dateto = g.val();
            }

            console.log("here daterange");
            console.log($scope.dataforsearch.datefrom);
            console.log($scope.dataforsearch.dateto);

            var ids = [];
            if (data1.type == 4) {
                _.each($scope.examplemodel, function (o2) {
                    var Row = o2.id;
                    ids.push(Row);
                })
            }
            else if (data1.type == 2) {
                _.each($scope.subexamplemodel, function (o2) {
                    var Row = o2.id;
                    ids.push(Row);
                })
            }
            else if (data1.type == 3) {
                _.each($scope.ssubexamplemodel, function (o2) {
                    var Row = o2.id;
                    ids.push(Row);
                })
            }
            else if (data1.type == 5) {
                _.each($scope.clusterModel, function (o2) {
                    var Row = o2.id;
                    ids.push(Row);
                })
            }

            var url = serviceBase + "api/NetProfit?datefrom=" + $scope.dataforsearch.datefrom + "&dateto=" + $scope.dataforsearch.dateto + "&type=" + data1.type + "&ids=" + ids;
            $http.get(url)
                .success(function (data) {
                    
                if (data.length == 0) {
                    alert("Data Not Present");   
                }
                else {
                    $scope.proft = data[0].value;
                    $scope.stores = data[0].netData;
                    alasql('SELECT CityName,HubName,SalesPersonName,SubcategoryName,SubsubcategoryName,SupplierName,itemname,Number,NetPurchasePrice,SellingPrice,QTY INTO XLSX("NetMargin.xlsx",{headers:true}) FROM ?', [$scope.stores]);
                   
                }
            });
        }
    
    //else {
    //    window.location = "#/404";
    //}
}]);